dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,293 @@
+# Yaml config design
+
+The router uses yaml configuration, and when creating new features or extending existing features you'll likely need to think about how configuration is exposed.
+
+In general users should have a pretty good idea of what a configuration option does without referring to the documentation.
+
+## Migrations
+
+We won't always get things right, and sometimes we'll need to provide [migrations](apollo-router/src/configuration/migrations/README.md) from old config to new config.
+
+Make sure you:
+1. Mention the change in the changelog
+2. Update docs
+3. Update any test configuration
+4. Create a migration test as detailed in [migrations](apollo-router/src/configuration/migrations/README.md)
+5. In your migration description tell the users what they have to update.
+
+## Process
+It should be obvious to the user what they are configuring and how it will affect Router behaviour. It's tricky for us as developers to know when something isn't obvious to users as often we are too close to the domain.
+
+Complex config should be run by the rest of the team. Ideally before writing any code, as coming to the team late can cause code churn and frustration. The process is as follows:
+1. In the github issue put the proposed config in.
+2. List any concerns.
+3. Notify the team that you are looking for request for comment.
+4. Schedule a meeting to discuss. (This is important, often design considerations will fall out of conversation)
```

Comment:
is this doc only for the router team, or would that be a reference for future contributors too? (that would have a harder time scheduling meetings with us)

Reply:
[14404db](https://github.com/apollographql/router/pull/2159/commits/14404db4744533ef697c7d4de62677ffb14c68d9)